zcash_client_sqlite/src/wallet.rs

```diff
@@ -1656,10 +1661,47 @@ fn subtree_scan_progress<P: consensus::Parameters>(
         // The outer option indicates whether or not we have recover-until height information;
         // the inner option indicates whether or not we were able to obtain a tree size given
         // the recover-until height.
-        let recover_until_size: Option<Option<u64>> = recover_until_height
-            // Find a tree size near to the recover-until height
-            .map(get_tree_size_near)
-            .transpose()?;
+        let recover_until_size: Option<Option<u64>> =
+            recover_until_height
+                .map(|h| {
+                    let size_from_blocks = stmt_start_tree_size_at
+                        .query_row(named_params![":start_height": u32::from(h)], |row| {
+                            row.get::<_, Option<u64>>(0)
+                        })
+                        .optional()?
+                        .flatten();
+
+                    match size_from_blocks {
+                        // We know the tree size at the start of the recover-until height.
+                        Some(size) => Ok::<_, SqliteClientError>(Some(size)),
+
+                        // If the recover-until height is equal to the chain tip height,
+                        // then this is almost certainly a newly-recovered wallet, and all
+                        // progress can count as recovery progress. Approximate the size
+                        // of the tree at the start of the block as equal to the size of
+                        // the tree at the end of the block; the scan progress will show
+                        // as 0/0 which is fine.
```

Comment:
This seems like a pretty narrow case to hit; shouldn't the linear extrapolation be able to handle this case? The only change that would be required would be to use a `map_or` for the result of the `checked_div`.

Also in general throwing an error that results from that `checked_div` seems unexpected - I think we would want to handle that error somehow, rather than propagate it.

Reply:
> This seems like a pretty narrow case to hit
You hit that case in the interval between creating a recover-from-seed wallet, and the first time the chain tip updates afterwards. That can be minutes of time; it's not rare.

> Also in general throwing an error that results from that checked_div seems unexpected
We're not, we're returning `Ok(None)` in that case, meaning we don't know the recover-until tree size.